libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -199,187 +228,214 @@ export class ConfigurationFile<TConfigurationFile> {
       | IConfigurationFileCacheEntry<TConfigurationFile>
       | undefined = this._configurationFileCache.get(resolvedConfigurationFilePath);
     if (!cacheEntry) {
-      try {
-        const resolvedConfigurationFilePathForErrors: string = ConfigurationFile._formatPathForError(
-          resolvedConfigurationFilePath
-        );
-
-        if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
-          throw new Error(
-            'A loop has been detected in the "extends" properties of configuration file at ' +
-              `"${resolvedConfigurationFilePathForErrors}".`
-          );
-        }
+      cacheEntry = await this._generateCacheEntryAsync(
+        resolvedConfigurationFilePath,
+        options,
+        visitedConfigurationFilePaths
+      );
+      this._configurationFileCache.set(resolvedConfigurationFilePath, cacheEntry);
+    }
 
-        visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
+    if (cacheEntry.parentFileError) {
+      throw cacheEntry.parentFileError;
+    } else if (cacheEntry.error) {
+      if (!isFirstFile || !options.ignoreIfNotExist || !FileSystem.isNotExistError(cacheEntry.error)) {
+        throw cacheEntry.error;
+      } else {
+        return undefined;
+      }
+    } else {
+      return cacheEntry.configurationFile;
+    }
+  }
 
-        let fileText: string;
-        try {
-          fileText = await FileSystem.readFileAsync(resolvedConfigurationFilePath);
-        } catch (e) {
-          if (FileSystem.isNotExistError(e)) {
-            e.message = `File does not exist: ${resolvedConfigurationFilePathForErrors}`;
-          }
+  private async _generateCacheEntryAsync(
+    resolvedConfigurationFilePath: string,
+    options: ILoadConfigurationFileOptions,
+    visitedConfigurationFilePaths: Set<string>
+  ): Promise<IConfigurationFileCacheEntry<TConfigurationFile>> {
+    try {
+      const resolvedConfigurationFilePathForErrors: string = ConfigurationFile._formatPathForError(
+        resolvedConfigurationFilePath
+      );
+
+      if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
+        throw new Error(
+          'A loop has been detected in the "extends" properties of configuration file at ' +
+            `"${resolvedConfigurationFilePathForErrors}".`
+        );
+      }
 
-          throw e;
-        }
+      visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
 
-        let configurationJson: IConfigurationJson & TConfigurationFile;
-        try {
-          configurationJson = await JsonFile.parseString(fileText);
-        } catch (e) {
-          throw new Error(`In config file "${resolvedConfigurationFilePathForErrors}": ${e}`);
+      let fileText: string;
+      try {
+        fileText = await FileSystem.readFileAsync(resolvedConfigurationFilePath);
+      } catch (e) {
+        if (FileSystem.isNotExistError(e)) {
+          e.message = `File does not exist: ${resolvedConfigurationFilePathForErrors}`;
```

Comment:
@iclanton Error objects should be considered as immutable once they are thrown. It doesn't seem like a good practice to modify an `Error` object after is caught, and then rethrow it.